### ISSUE TITLE
No exception thrown in catch-block if finally-block contains a return statement 

### ISSUE THREAD

**Issue by PhilippeMaincon** [JuliaLang/julia#28421]
The following MWE demonstrates a bug of mine: line 14 assume any exception has a field .msg .  This throws an exception, but in the presence of a return statement in the finally block, the exception is not propagated, and this was hard to track down.

I am of the opinion that, to help code development, an exception arising in a catch statement should be thrown, bypassing finally.  This will leave files unclosed, but help correcting the catch statement. Or the warning should be in big font in the manual!

`function bungle(c)
    out = 0
    try
        @printf "\n\n\n\Bungling away\n"
        out = 1
        if c == 1
            @assert false @sprintf "We have a main %s undervolt" "B"
        elseif c==2
            a = sqrt(-1.)
        end
    catch ex
        out = 2
        @printf "catch block\n"
        @printf "Houston we have a problem: %s\n" ex.msg
    finally
        @printf "Shutting down.\n"
        return out
    end
end 
@show bungle(0)
@show bungle(1)
@show bungle(2)`

**Comment by PhilippeMaincon** [comment]
Cf. discussion on [https://discourse.julialang.org/t/failure-to-catch-an-exception/12822/10](url)

**Comment by bkamins** [comment]
The issue is more general. Also without `catch` block, if `finally` contains return statement then any uncaught exception (no matter if it is thrown in `try` or `catch`) will not be propagated.

**Comment by bkamins** [comment]
An additional set of related cases:

```
function bar1()
    try
    bar1()
    catch
        println("caught")
    end
end

function bar2()
    try
    bar2()
    finally
        println("finalized")
    end
end

function bar3()
    try
    bar3()
    finally
        return "done"
    end
end
```

The output is:

* running `bar1()` segfaults;
* running `bar2()` segfaults;
* running `bar3()` is unpredictable: on the first run it returns `"done"` without any error, but on the second run it segfaults;

**Comment by bkamins** [comment]
Just to add if I remove `println` in `bar1` and `bar2` the behavior changes:

```
function bar1()
    try
        bar1()
    catch
        1
    end
end

function bar2()
    try
        bar2()
    finally
        1
    end
end
```

And now:
* `bar1()` hangs Julia 0.7-beta2 (i.e. no segfault, but CPU usage is 100% and process waits), but it properly returned `1` on Julia 0.6
* `bar2()` properly propagates `StackOverflowError`

**Comment by StefanKarpinski** [comment]
No matter how nonsensical, it should still work.

**Comment by StefanKarpinski** [comment]
@Keno I assume this is a bug with the new optimizer?

**Comment by bkamins** [comment]
There are several issues at work here. One is with the difference between 0.6 and 0.7 and segfaults. And I guess this should be fixed no matter what we decide about functionality. Right?

The other is how `return` in `finally` should be handled. I guess we have three options (all are potentially sensible and I might not see some consequences in corner cases):
1. stay with the current behavior (i.e. it silently hides uncaught exception) - then we should clearly document it
2. make sure that even with `return` in `finally` the exception propagates up; this probably requires changes in how `try-catch-finally` is lowered and is breaking
3. disallow `return` in `finally` (I guess this is what @PhilippeMaincon suggested) - this is also breaking

I am not sure which of them if preferred by core devs (for sure option 1 is cheapest to implement 😄).

**Comment by StefanKarpinski** [comment]
This is a bug, plain and simple. Presence of a return in a finally block should not supress exceptions.

**Comment by Keno** [comment]
As for the segfaults, certain c libraries (including libuv and LLVM) really do not like to be called at the edge of the stack (except perhaps reserve a couple pages for emergency stack - and then if we happen to be in a ccall allow those couple extra pages).

**Comment by Keno** [comment]
> @Keno I assume this is a bug with the new optimizer?

I don't think so. AFAIK, the lowering has always been to rethrow the exception after the finally block. The difference between 0.6 and 0.7 is due to whether or not the code in the catch block throws.

**Comment by JeffBezanson** [comment]
The stack overflow cases are NOT related and that is really muddying the waters here. Also they "work" for me on linux, @bkamins what platform are you on? Stack overflow continues to be unreliable on mac os.

`return`ing from a finally block definitely doesn't seem like a good idea, but from what I can tell our behavior here matches java and python.

**Comment by bkamins** [comment]
Regarding the core issue: that is why I have asked. I confirm that current behavior matches Python and Java (and we probably can, but do not have to match it).

Actually I think that Java specification of try-catch-finally is a good reference of a complete write-up of the behavior. In this sense `return` statement in `finally` can be classified as *abrupt completion*:

>  If the finally block completes abruptly for reason S , then the try statement completes abruptly for reason S

in Java. Anyway - if we stick with it it would be good to document it.

Regarding `StackOverflowException` - I have used it as it is typically a corner case for testing. Now - the results I reported are on Win10. I have checked Linux and things look better. `bar3` behaves correctly, `bar1` and `bar2` have problems only when they try to print something. From @Keno said I understand that it will be hard to control, so we have to accept those cases as problemetic.

**Comment by oscardssmith** [comment]
I think we should error here. We should never take design cues from Java, and we tend to be less permissive than python. C# makes it an explicit error because it is something you probably never should do.

**Comment by Keno** [comment]
I think a syntax error for `return` inside `finally` would be fine. We already disallow things like arbitrary gotos across `try/catch` blocks.

**Comment by JeffBezanson** [comment]
Yes, I agree it would be reasonable to make this an error. Of course the timing of this is bad; I certainly wouldn't want to put more breaking changes in 0.7.

**Comment by bkamins** [comment]
Whatever decision is made about the functionality and timing I think it is important to take care of a case of exception thrown in finally block as this is something that we are not able to disallow and it has the same consequence as `return` (at least currently).

**Comment by JeffBezanson** [comment]
If an exception is thrown in a finally block, then yes it hides the previous exception, but the new exception still propagates so you still at least see the last error. `return` is different because it hides the last exception completely.

**Comment by StefanKarpinski** [comment]
I've written this down in my list of unstable behaviors that we may still change. I think we could reasonable deprecate return within finally in the 1.x series given that it very likely doesn't do what half of the people would expect it to do.

**Comment by vtjnash** [comment]
Both exceptions now propagate from the finally block, though still not if the exception happened in the catch block--then it is simply discarded by the return from the finally block
```
julia> try
           error("try error")
       finally
           sqrt(-10)
       end
ERROR: DomainError with -10.0:
sqrt was called with a negative real argument but will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).
Stacktrace:
 [1] throw_complex_domainerror(f::Symbol, x::Float64)
   @ Base.Math ./math.jl:33
 [2] sqrt
   @ ./math.jl:691 [inlined]
 [3] sqrt(x::Int64)
   @ Base.Math ./math.jl:1586
 [4] top-level scope
   @ REPL[3]:4

caused by: try error
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] top-level scope
   @ REPL[3]:2
```
vs
```
julia> try
            error("try error")
        catch
            sqrt(-10)
        finally
        end
ERROR: DomainError with -10.0:
sqrt was called with a negative real argument but will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).
Stacktrace:
 [1] throw_complex_domainerror(f::Symbol, x::Float64)
   @ Base.Math ./math.jl:33
 [2] sqrt
   @ ./math.jl:691 [inlined]
 [3] sqrt(x::Int64)
   @ Base.Math ./math.jl:1586
 [4] top-level scope
   @ REPL[1]:4

caused by: try error
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] top-level scope
   @ REPL[1]:2                      
```
vs
```
julia> try
            error("try error")
        catch
            sqrt(-10)
        finally
        return
        end
#nothing printed here
```


Note that the `return` also corrupts the error stack, so that *all* future exceptions are broken as well:
```
julia> error("bad")
ERROR: bad
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] top-level scope
   @ REPL[2]:1

caused by: DomainError with -10.0:
sqrt was called with a negative real argument but will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).
Stacktrace:
 [1] throw_complex_domainerror(f::Symbol, x::Float64)
   @ Base.Math ./math.jl:33
 [2] sqrt
   @ ./math.jl:691 [inlined]
 [3] sqrt(x::Int64)
   @ Base.Math ./math.jl:1586
 [4] top-level scope
   @ REPL[1]:4

caused by: try error
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] top-level scope
   @ REPL[1]:2
```

**Comment by shamsulalam1114** [comment]
function bungle(c)
    out = 0
    try
        @printf "\n\n\nBungling away\n"
        out = 1
        if c == 1
            @assert false @sprintf "We have a main %s undervolt" "B"
        elseif c==2
            a = sqrt(-1.)
        end
    catch ex
        out = 2
        @printf "catch block\n"
        @printf "Houston we have a problem: %s\n" (haskey(ex, :msg) ? ex.msg : "no message")
        rethrow(ex)
    finally
        @printf "Shutting down.\n"
    end
    return out
end 

@show bungle(0)
@show bungle(1)
@show bungle(2)
any exception in the catch block will indeed be thrown, bypassing the finally block. Be sure to test it out and see if it behaves as expected!

By the way, juggling between error handling and resource management can be like walking on a tightrope. You get better each time you cross it!
